Cache results of ResolveTargetingPackAssets across builds

rainersigwald opened the pull request:
This PR uses MSBuild's RegisteredTaskObject machinery to cache (for the lifetime of the process) the results of resolving targeting pack assets. This task runs on every build because it must produce items to populate `@(Reference)`, but its output basically never changes since the targeting packs are static.

The caching is two-layered:

1. Cache the parsing of the targeting pack (`FrameworkList.xml` and friends) so that those files need to be read only once per MSBuild worker node per build.
2. Cache the full results of the task, including deduplication in the case of multiple targeting packs.

Fixes #19282.

I tested this with a no-op incremental build scenario (of https://github.com/OrchardCMS/OrchardCore/commit/ec2154ae099913ee81751f30c4f7660952fe63c9). Worker nodes were reused for the benchmarks.

Filtering to just this task, you can see a significant reduction in total time spent:

![image](https://user-images.githubusercontent.com/3347530/128575548-1dce1afe-3ae2-4ab4-9bee-310d2bed09b0.png)

The overall time is improved but not dramatically (on the order of 2%):

| Command | Mean [s] | Min [s] | Max [s] | Relative |
|:---|---:|---:|---:|---:|
| `S:\benchmarkscripts\orchardcore\Build.bat S:\sdk\sdk-main` | 22.056 ± 1.190 | 20.107 | 23.662 | 1.02 ± 0.09 |
| `S:\benchmarkscripts\orchardcore\Build.bat S:\sdk\cacheharder` | 21.725 ± 1.508 | 19.727 | 24.428 | 1.00 |

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -46,6 +46,8 @@ public class ResolveTargetingPackAssets : TaskBase
         [Output]
         public ITaskItem[] UsedRuntimeFrameworks { get; set; }
 
+        private static readonly bool s_AllowCacheLookup = System.Environment.GetEnvironmentVariable("DOTNET_SDK_ALLOWTARGETINGPACKCACHING") != "0";
```
It doesn't look like there are a lot of SDK escape hatches like this. Does this look ok?

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -208,10 +266,37 @@ private void AddNetStandardTargetingPackAssets(ITaskItem targetingPack, string t
+                  RegisteredTaskObjectLifetime.AppDomain /* TODO: should this be build? */) 
+                is FrameworkListCacheEntry cacheEntry)
+            {
+                // TODO: probe files to see if they've been modified? or just assume not
```
The meta-question here is: is there any supported scenario where the contents of the targeting pack will change? Or can we cache it forever?

rainersigwald replied:
Discussed this with @dsplaisted offline and we decided "process lifetime" is probably ok, as we don't expect targeting packs to change.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -208,10 +266,37 @@ private void AddNetStandardTargetingPackAssets(ITaskItem targetingPack, string t
+            {
+                // TODO: probe files to see if they've been modified? or just assume not
+
+                // TODO: is this safe? do we need to clone them or something?
```
This is on me and I now suspect the answer is "it's fine".

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -61,6 +63,47 @@ protected override void ExecuteCore()
+            {
+                // TODO: probe files to see if they've been modified? or just assume not
+
+                // TODO: is this safe? do we need to clone them or something?
```
A task can safely cache `TaskItem`s as far as I can tell: dotnet/msbuild#6754

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -208,13 +267,43 @@ private void AddNetStandardTargetingPackAssets(ITaskItem targetingPack, string t
-
             string profile = targetingPack.GetMetadata("Profile");
 
+            StringBuilder hashKey = new(nameof(ResolveTargetingPackAssets) + nameof(AddItemsFromFrameworkList));
```
Same feedback here about making sure we're caching all the inputs and adding test coverage for that if possible.

rainersigwald replied:
This one is harder since it's internal but I'll see what I can do.

rainersigwald replied:
It was indeed harder but a bit of refactoring made it work ok. Look good now?

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -61,6 +63,47 @@ protected override void ExecuteCore()
+                hashKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.PackageConflictPreferredPackages));
+            }
+
+            string hashKey = hashKeyBuilder.ToString();
```
That was my inclination as well but I can turn it off if anyone feels strongly.

rainersigwald commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -61,6 +63,47 @@ protected override void ExecuteCore()
             var resolvedTargetingPacks = ResolvedTargetingPacks.ToDictionary(item => item.ItemSpec, StringComparer.OrdinalIgnoreCase);
 
+            StringBuilder hashKeyBuilder = new(nameof(ResolveTargetingPackAssets) + nameof(ExecuteCore));
+            hashKeyBuilder.AppendLine();
```
That test is slick and I love it. I'll adopt one much like it.

rainersigwald replied:
Done.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/ResolveTargetingPackAssetsTests.cs:
```diff
@@ -12,15 +16,82 @@ public class ResolveTargetingPackAssetsTests
+            ((MockBuildEngine4)task.BuildEngine).RegisteredTaskObjects.Count.Should().Be(2,
+                because: "there should be a cache entry for the overall lookup and for the specific targeting pack");
+
+            task.Execute().Should().BeTrue();
```
Should we be creating a new task instance here instead of reusing the existing one?

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -162,6 +190,69 @@ protected override void ExecuteCore()
+            foreach (var resolvedTargetingPack in ResolvedTargetingPacks)
+            {
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.ItemSpec);
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.Path));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata("TargetingPackFormat"));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata("TargetFramework"));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.NuGetPackageId));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.NuGetPackageVersion));
+                cacheKeyBuilder.AppendLine(resolvedTargetingPack.GetMetadata(MetadataKeys.PackageConflictPreferredPackages));
```
Ideally the tests would verify that the key depends on any metadata items that are read.  The way to do that would probably be to use a strongly-typed wrapper for all metadata access.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks/ResolveTargetingPackAssets.cs:
```diff
@@ -208,13 +299,34 @@ private void AddNetStandardTargetingPackAssets(ITaskItem targetingPack, string t
             string targetingPackDllFolder,
             ITaskItem targetingPack, List<TaskItem> referenceItems, List<TaskItem> analyzerItems)
         {
```
Can this be a static method that takes a `FrameworkListDefinition` as an argument?  Then it won't be able to accidentally access data that's not included in the key.

dsplaisted reviewed (approved):
Looks good.  I think it would be nice to also cover the item metadata in the tests.

rainersigwald committed:
Cache items parsed out of targeting packs

Uses MSBuild's RegisteredTaskObject machinery to cache (for the lifetime
of the process) the results of resolving targeting pack assets. This
task runs on every build because it must produce items to populate
@(Reference), but its output basically never changes since the targeting
packs are static.

The caching is two-layered:

1. Cache the parsing of the targeting pack (FrameworkList.xml and
   friends) so that those files need to be read only once per MSBuild
   worker node per build.
2. Cache the full results of the task, including deduplication in the
   case of multiple targeting packs.

Fixes #19282.

rainersigwald commented:
Test failure was completely legit: https://github.com/dotnet/sdk/compare/4d159ab1d7e8d0560509e8176f1127eac299de3d..8fb8739032f205dc88cfdb66bb95f828ef84e17a

I will add a smaller UT for this task to catch that too.

rainersigwald committed:
Add test for UsedRuntimeFrameworks

rainersigwald commented:
Updated individual-task results (OrchardCore, 24 build nodes):

![image](https://user-images.githubusercontent.com/3347530/131858905-7a033965-9c63-4aad-b3b4-a237c22c27e8.png)

Overall time is much noisier. I suspect my initial 1-2% improvement was noise there. But I can't see a reason why speeding up this task would _increase_ overall time.

dsplaisted commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolveTargetingPackAssetsTask.cs:
```diff
@@ -0,0 +1,424 @@
+
+        private ResolveTargetingPackAssets InitializeTask(out string mockPackageDirectory)
+        {
+            mockPackageDirectory = Path.Combine(Path.GetTempPath(), "dotnetSdkTests", Path.GetRandomFileName());
```
I'm not a fan of putting test state in temporary directories that don't get cleaned up and are hard to find for debugging.  Can we make this test class derive from `SdkTest` and then use `_testAssetsManager.CreateTestDirectory()` or something?

Most of the task unit tests don't derive from `SdkTest`, but `GivenAResolveRuntimePackAssetsTask` does, so it should be OK.

rainersigwald replied:
This was preexisting but sure.

rainersigwald replied:
Is 289ee935ca what you were thinking of?

dsplaisted replied:
Pretty much, but I think you'll need to update it to use a different directory each time it's called as part of the same test.  We generally do this with an `identifier` parameter that you can pass through to CreateTestDirectory.

When running in CI, we will fail if it tries to create a test directory that already exists.  This helps prevent flakiness where different tests end up using the same test directory.  We don't do that locally, since otherwise every time you re-ran a test it would fail.

rainersigwald replied:
Done in 5f64f82fa9 but right now no test calls it twice so I didn't actually call it. Do you want me to pass it in in every call anyway?

dsplaisted replied:
OK, I got confused by the different overloads of `InitializeTask`: https://github.com/dotnet/sdk/blob/5f64f82fa9faf86654b837be4c0b695e64f4caf2/src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolveTargetingPackAssetsTask.cs#L74-L87

If no test actually calls it twice then you don't even need to take the identifier as an argument, I just thought the tests were re-creating the same directory and that was going to cause failures in CI.

rainersigwald replied:
Pushed a revert + some strategic renames to hopefully clarify for the future.

dsplaisted replied:
Thanks.  I already signed off so feel free to merge.

dsplaisted reviewed (approved):
Great job, this is really cool!

AraHaan commented:
Anyone able to test if this can also speed up say “compiling roslyn” itself or say compiling everything in dotnet/runtime locally.

rainersigwald committed:
Switch to SdkTest

rainersigwald commented:
> Anyone able to test if this can also speed up say “compiling roslyn” itself or say compiling everything in dotnet/runtime locally.

@AraHaan I'd expect it to speed it up but not significantly at those scales. This task isn't a huge proportion of build time, just easy to fix.

rainersigwald committed:
Allow unique identifier

rainersigwald committed:
Revert "Allow unique identifier"

This reverts commit 5f64f82fa9faf86654b837be4c0b695e64f4caf2.

rainersigwald committed:
Disambiguate InitializeTask calls

AraHaan commented on src/Tasks/Microsoft.NET.Build.Tasks.UnitTests/GivenAResolveTargetingPackAssetsTask.cs:
```diff
@@ -125,10 +125,9 @@ public void Given_Passing_ResolvedTargetingPacks_A_Different_Language_Parses_Aga
 
         private ResolveTargetingPackAssets InitializeTask(out string mockPackageDirectory,
-            [CallerMemberName] string testName = nameof(GivenAResolvePackageAssetsTask),
-            string identifier = null)
```
Why did this get reverted?

rainersigwald replied:
It's not needed today: https://github.com/dotnet/sdk/pull/19570#discussion_r703831303
